empyrical/utils.py

```diff
@@ -143,17 +142,17 @@ def down(returns, factor_returns, **kwargs):
 
 def _roll_ndarray(func, window, *args, **kwargs):
     data = []
-    for i in range(window, len(args[0])):
```

Comment:
I wouldn't couple it to this PR, but in the future we could probably dramatically speed up the numpy case by using something like [`zipline.utils.numpy_utils.rolling_window`](https://github.com/quantopian/zipline/blob/master/zipline/utils/numpy_utils.py#L253-L319) to implement these as aggregations on a restrided array, which would remove the python overhead of slicing on each iteration.

For example, using `np.mean`, I see about 100x speedup from this:

```python
In [28]: %timeit roll(data, function=np.mean, window=50)
100 loops, best of 3: 3.83 ms per loop

In [29]: %timeit rolling_window(data, 50).mean(axis=1)
10000 loops, best of 3: 35.5 µs per loop
```

For the pandas case, we can just use pandas' native `.rolling()` methods for the single data-input functions.  For the multi-argument functions we'd have to be a little more clever.

Reply:
Unfortunately `rolling()` does not allow for different subsampling and window-lengths. At least not that I could figure out.

Reply:
@ssanderson I can't say I disagree with your style and readability points; there are certainly some areas of the code that could be smartened up. I was going for minimal code diff for this PR; focusing purely on correcting what is there in a way that was timely and easy to review. 

Regarding `rolling_window` and pandas' native `rolling`, I'd be in favor of porting over some of the rolling methods from the zipline area you referenced rather than adding all of `zipline` as a dependency, which may conflict with the desire to keep this repo fairly lean and dependency light.

To date, I've found pandas' `rolling()` to lack some of the flexibility we wanted here, but it has been a minute since I've looked into it deeply.

I'm happy to take another crack at a long-term fix with you guys, but I've got a pretty stacked couple of weeks ahead of me. Perhaps merge this as a short-term fix and add an issue for some style and speed optimization in a few weeks ?

Reply:
> I was going for minimal code diff for this PR; focusing purely on correcting what is there in a way that was timely and easy to review.

Yup! I didn't mean to imply that this any of this should be a blocker for fixing the bug.  Just wanted to note that this kind of code is tricky (both for correctness and performance) to write "by hand", so if we can leverage "native" tooling for doing rolling computations it's probably a good idea to do so.

Reply:
> I'm happy to take another crack at a long-term fix with you guys, but I've got a pretty stacked couple of weeks ahead of me. Perhaps merge this as a short-term fix and add an issue for some style and speed optimization in a few weeks ?

Definitely fine with merging this as is.  re: future improvements, @llllllllll is currently working on overhauling how Zipline/Quantopian manage risk metrics (see https://github.com/quantopian/zipline/pull/2081), and part of that work is converting many of our risk metrics into vectorized rolling computations at the end of backtests instead of streaming computations during backtests, so I expect that he might take a pass at this in the next few weeks as well.